src/middlewared/middlewared/plugins/network.py

```diff
@@ -352,6 +353,102 @@ def iface_extend(self, iface_state, configs, ha_hardware, fake=False):
+        any type of warning.
+
+        NOTE: This makes 2 assumptions
+        1. interface.create/update/delete must have been called before
+            calling this method
+        2. this method must be called before `interface.sync` is called
+
+        This method exists for the predominant scenario for new users...
+        1. fresh install SCALE
+        2. all interfaces start DHCPv4 (v6 is ignored for now)
+        3. 1 of the interfaces receives an IP address
+        4. along with the IP, the kernel receives a default route
+            (by design, of course)
+        5. user goes to configure this interface as having a static
+            IP address
+        6. as we go through and "sync" the changes, we remove the default
+            route because it exists in the kernel FIB but doesn't exist
+            in the database.
+        7. IF the user is connecting via layer3, then they will lose all
+            access to the TrueNAS and never be able to finalize the changes
+            to the network because we ripped out the default route which
+            is how they were communicating to begin with.
+
+        In the above scenario, we're going to try and prevent this by doing
+        the following:
+        1. fresh install SCALE
+        2. all interfaces start DHCPv4
+        3. default route is received
+        4. user configures an interface
+        5. When user pushes "Test Changes" (interface.sync), webUI will call
+            network.configuration.default_route_will_be_removed BEFORE interface.sync
+        6. if network.configuration.default_route_will_be_removed returns True,
+            then webUI will open a new modal dialog that gives the end-user
+            ample warning/verbiage describing the situation. Furthermore, the
+            modal will allow the user to input a default gateway
+        7. if user gives gateway, webUI will call this method providing the info
+            and we'll validate accordingly
+        8. OR if user doesn't give gateway, they will need to "confirm" this is
+            desired
+        9. the default gateway provided to us (if given by end-user) will be stored
+            in the same in-memory cache that we use for storing the interface changes
+            and will be rolledback accordingly in this plugin just like everything else
+
+        There are a few other scenarios where this is beneficial, but the one listed above
+        is seen most often by end-users/support team.
+        """
+        if not self._original_datastores:
+            raise CallError('There are no pending interface changes.')
+
+        gw = ip_address(gw)
+        defgw = {'gc_ipv4gateway': gw.exploded}
+        for iface in await self.middleware.call('datastore.query', 'network.interfaces'):
+            if gw in ip_interface(f'{iface["int_address"]}/{iface["int_netmask"]}').network:
+                await self.middleware.call('datastore.update', 'network.globalconfiguration', 1, defgw)
```

Comment:
Are we sure this row always exists and has `id = 1`? I have never made a good assumption about our database that wasn't soon broken by a ticket from a user that uses FreeNAS since 9.1.

Reply:
I'm willing to make that assumption since no row in that table will cause other things to break before we get here and if `id != 1` then I'd like to know how they got into that state 😄